File: functions/models/competitions/competition.js
/* eslint-disable*/
class Competition {
    constructor(
        competitionId,
        storeFK,
        townOrInstitution,
        dateTime,
        joiningFee,
        numberOfGrandPrices,
        pickingMultipleInSeconds,
        timeBetweenPricePickingAndGroupPicking,
        displayPeriodAfterWinners
    ) {
        this.competitionId = competitionId,
            this.storeFK = storeFK,
            this.townOrInstitution = townOrInstitution,
            this.isLive = true,
            this.dateTime = dateTime,
            this.joiningFee = joiningFee,
            this.numberOfGrandPrices = numberOfGrandPrices,
            this.isOver = false,
            this.grandPricesGridId = "-",
            this.competitorsGridId = "-",
            this.groupPickingStartTime = -1,
            this.pickingMultipleInSeconds = pickingMultipleInSeconds,
            this.timeBetweenPricePickingAndGroupPicking = timeBetweenPricePickingAndGroupPicking,
            this.displayPeriodAfterWinners = displayPeriodAfterWinners, // must switch to 5 minute (300)

            this.grandPricesOrder = [],
            this.isWonGrandPricePicked = false,
            this.competitorsOrder = [],
            this.isWonCompetitorGroupPicked = false,
            this.competitionState = "on-count-down",
            this.wonPrice = null,
            this.wonGroup = null
    }

    toJson() {
        return {
            competitionId: this.competitionId,
            storeFK: this.storeFK,
            townOrInstitution: this.townOrInstitution,
            isLive: this.isLive,
            dateTime: this.dateTime,
            joiningFee: this.joiningFee,
            numberOfGrandPrices: this.numberOfGrandPrices,
            isOver: this.isOver,
            grandPricesGridId: this.grandPricesGridId,
            competitorsGridId: this.competitorsGridId,
            groupPickingStartTime: this.groupPickingStartTime,
            pickingMultipleInSeconds: this.pickingMultipleInSeconds,
            timeBetweenPricePickingAndGroupPicking: this.timeBetweenPricePickingAndGroupPicking,
            displayPeriodAfterWinners: this.displayPeriodAfterWinners, // must switch to 5 minute (300)

            grandPricesOrder: this.grandPricesOrder,
            isWonGrandPricePicked: this.isWonGrandPricePicked,
            competitorsOrder: this.competitorsOrder,
            isWonCompetitorGroupPicked: this.isWonCompetitorGroupPicked,
            competitionState: this.competitionState,
            wonPrice: this.wonPrice.toJson(),
            wonGroup: this.wonGroup.toJson(),

        };
    }


}

File: lib/models/competitions/competition.dart
import '../locations/converter.dart';
import '../locations/supported_town_or_institution.dart';
import '/models/hosting areas/draw_grand_price.dart';

import '../users/group.dart';
import 'competition_state.dart';

import 'dart:developer' as debug;

// Collection Name /competition/competitionId
// Branch : competition_resources_crud ->  create_competition_resources_front_end
class Competition {
  String? competitionId;
  String hostingAreaFK;

  bool isLive;
  bool isOver;
  int joiningFee;

  SupportedTownOrInstitution competitionTownOrInstitution;

  DateTime dateTime;
  int pickingMultipleInSeconds;

  String? grandPricesGridId;
  List<int> grandPricesOrder;
  int numberOfGrandPrices;

  String? competitorsGridId;
  List<String> competitorsOrder;
  int groupPickingStartTime;

  int? timeBetweenPricePickingAndGroupPicking;
  int? displayPeriodAfterWinners;

  bool? isWonGrandPricePicked;
  bool? isWonCompetitorGroupPicked;

  CompetitionState? competitionState;

  DrawGrandPrice? wonPrice;
  Group? wonGroup;

  Competition({
    this.competitionId,
    required this.hostingAreaFK,
    required this.competitionTownOrInstitution,
    this.isLive = false,
    required this.dateTime,
    required this.joiningFee,
    this.isOver = false,
    required this.grandPricesOrder,
    this.isWonGrandPricePicked = false,
    this.grandPricesGridId,
    required this.numberOfGrandPrices,
    required this.competitorsOrder,
    this.isWonCompetitorGroupPicked = false,
    this.competitorsGridId,
    required this.pickingMultipleInSeconds,
    this.timeBetweenPricePickingAndGroupPicking = 5,
    required this.groupPickingStartTime,
    this.displayPeriodAfterWinners = 30,
    this.competitionState,
    this.wonGroup,
    this.wonPrice,
  });

  factory Competition.fromJson(dynamic json) {
    List<int> grandPricesOrder =
        convertGrandPricesOrder(json["grandPricesOrder"]);
    List<String> competitorsOrder =
        convertCompetitorsOrder(json['competitorsOrder']);
    int pickingMultipleInSeconds = json['pickingMultipleInSeconds'];
    int timeBetweenPricePickingAndGroupPicking =
        json["timeBetweenPricePickingAndGroupPicking"];

    return Competition(
        wonPrice: json['wonPrice'] != null
            ? DrawGrandPrice.fromJson(json['wonPrice'])
            : null,
        wonGroup:
            json['wonGroup'] != null ? Group.fromJson(json['wonGroup']) : null,
        pickingMultipleInSeconds: pickingMultipleInSeconds,
        timeBetweenPricePickingAndGroupPicking:
            timeBetweenPricePickingAndGroupPicking,
        groupPickingStartTime: json['groupPickingStartTime'],
        displayPeriodAfterWinners: json['displayPeriodAfterWinners'],
        competitionId: json['competitionId'],
        hostingAreaFK: json['hostingAreaFK'],
        competitionTownOrInstitution: SupportedTownOrInstitution.fromJson(
            json['competitionTownOrInstitution']),
        isLive: json['isLive'],
        dateTime: DateTime(
          json['dateTime']['year'],
          json['dateTime']['month'],
          json['dateTime']['day'],
          json['dateTime']['hour'],
          json['dateTime']['minute'],
        ),
        joiningFee: json['joiningFee'],
        isOver: json['isOver'],
        grandPricesOrder: grandPricesOrder,
        isWonGrandPricePicked: json['isWonGrandPricePicked'],
        grandPricesGridId: json['grandPricesGridId'],
        numberOfGrandPrices: json["numberOfGrandPrices"],
        competitorsOrder: competitorsOrder,
        isWonCompetitorGroupPicked: json['isWonCompetitorGroupPicked'],
        competitorsGridId: json['competitorsGridId'],
        competitionState:
            Converter.toCompetitionState(json['competitionState']));
  }

  void setIsOver(bool isOver) {
    this.isOver = isOver;
  }

  void setIsWonPricePicked(bool isWonCompetitorGroupPicked) {
    this.isWonCompetitorGroupPicked = isWonCompetitorGroupPicked;
  }

  static List<int> convertGrandPricesOrder(List<dynamic> order) {
    List<int> list = [];

    for (int i = 0; i < order.length; i++) {
      list.add(order[i]);
    }

    return list;
  }

  static List<String> convertCompetitorsOrder(List<dynamic> order) {
    List<String> list = [];

    for (int i = 0; i < order.length; i++) {
      list.add(order[i]);
    }

    return list;
  }
}
